tests/test_base.py

```diff
@@ -0,0 +1,941 @@
+# -*- coding: utf-8 -*-
+import datetime
+import enum
+import re
+
+import pytest
+from pytz import utc
+
+from cleancat import (
+    Bool, Choices, DateTime, Dict, Email, Embedded, Enum, Field, Integer,
+    List, Regex, RelaxedURL, Schema, SortedSet, StopValidation, String,
+    TrimmedString, URL, ValidationError
+)
+
+
+class TestField:
+
+    def test_it_supports_multiple_base_types(self):
+        class IntOrStrField(Field):
+            base_type = (int, str)
+
+        assert IntOrStrField().clean(5) == 5
+        assert IntOrStrField().clean('five') == 'five'
+
+        with pytest.raises(ValidationError) as e:
+            assert IntOrStrField().clean(4.5)
+        assert e.value.args[0] == 'Value must be of int or str type.'
+
+
+class TestStringField:
+
+    def test_it_accepts_valid_input(self):
+        value = 'hello world'
+        assert String().clean(value) == value
+
+    @pytest.mark.parametrize('value', ['', None])
+    def test_it_enforces_the_required_flag(self, value):
+        with pytest.raises(ValidationError) as e:
+            String().clean(value)
+        assert e.value.args[0] == 'This field is required.'
+
+    def test_it_accepts_valid_input_if_not_required(self):
+        value = 'hello world'
+        assert String().clean(value) == value
+
+    @pytest.mark.parametrize('value', ['', None])
+    def test_it_can_be_optional(self, value):
+        with pytest.raises(StopValidation) as e:
+            String(required=False).clean(value)
+        assert e.value.args[0] == ''
+
+    def test_it_enforces_valid_data_type(self):
+        with pytest.raises(ValidationError) as e:
+            String().clean(True)
+        assert e.value.args[0] in (
+            # TODO this should be consistent and say the same thing for both.
+            'Value must be of basestring type.',  # Py2
+            'Value must be of str type.',  # Py3
+        )
+
+    @pytest.mark.parametrize('value,valid', [
+        ('long enough', True),
+        ('short', False),
+    ])
+    def test_it_enforces_min_length(self, value, valid):
+        field = String(min_length=10)
+        if valid:
+            assert field.clean(value) == value
+        else:
+            with pytest.raises(ValidationError) as e:
+                field.clean(value)
+            assert e.value.args[0] == (
```

Comment:
`with pytest.raises(..., message='The value must be ...')`

Reply:
[Docs for `pytest.raises`](https://docs.pytest.org/en/latest/reference.html#pytest-raises) say:

> if specified, provides a custom failure message if the exception is not raised

It's not to assert what the exception's message is.

Reply:
I guess `match` is what we could use. That said, I don't think it provides much value over this assert.

Reply:
Ah yep, I mixed up keywords. `match` it is.

Reply:
Ah yep, I mixed up keywords. `match` it is.

Reply:
Ah yep, I mixed up keywords. `match` it is.

Reply:
Ah yep, I mixed up keywords. `match` it is.

Reply:
Done in 5027de2